File: test/util/PriorityQueueTest.java
package util;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class PriorityQueueTest {

    private PriorityQueue<String> queue;

    @Before
    public void setUp() throws Exception {
        queue = new PriorityQueue<>();
    }

    @Test
    public void newQueueIsEmpty() throws Exception {
        assertTrue(queue.isEmpty());
    }

    @Test
    public void newQueueHasSizeZero() {
        assertEquals(0, queue.size());
    }

    @Test
    public void addGivesSize1() {
        queue.push(1, "hello");
        assertEquals(1, queue.size());
    }

    @Test
    public void addTwoGivesSize2() {
        queue.push(2,"hello");
        queue.push(3, "world");

        assertEquals(2, queue.size());
    }

    @Test
    public void addBubblesUp() {
        queue.push(3,"hello");
        queue.push(2, "world");

        assertEquals("world", queue.peek());
    }

    @Test
    public void addBubblesUp2() {
        queue.push(3,"hello");
        queue.push(2, "world");
        queue.push(5, "dag");
        queue.push(1, "wereld");
        queue.push(10, "hoi");

        assertEquals("wereld", queue.peek());
    }

    @Test
    public void popBubblesDown() {
        queue.push(3,"hello");
        queue.push(2, "world");
        queue.push(5, "dag");
        queue.push(1, "wereld");
        queue.push(10, "oom");

        queue.pop();
        assertEquals("world", queue.peek());
    }

    @Test
    public void popBubblesDown2() {
        queue.push(20, "word");
        queue.push(0, "test");
        queue.push(3,"hello");
        queue.push(2, "world");
        queue.push(5, "dag");
        queue.push(1, "wereld");
        queue.push(10, "oom");

        assertEquals("test", queue.pop());
        assertEquals("wereld", queue.pop());
        assertEquals("world", queue.pop());
        assertEquals("hello", queue.pop());
        assertEquals("dag", queue.pop());
        assertEquals("oom", queue.pop());
        assertEquals("word", queue.pop());
    }
}

File: Go/util/PriorityQueue_test.go
package util

import (
	"fmt"
)

func ExampleQueueIsEmpty() {
	pq := New()
	fmt.Println(pq.IsEmpty())
	// Output: true
}

func ExampleQueueHasSizeZero() {
	pq := New()
	fmt.Println(pq.Size())
	// Output: 0
}

func ExamplePushGivesSizeOne() {
	pq := New()
	pq.Push(10, "hello world")
	fmt.Println(pq.Size())
	// Output: 1
}

func ExamplePushTwoGivesSizeTwo() {
	pq := New()
	pq.Push(10, "hello")
	pq.Push(200, "world")
	fmt.Println(pq.Size())
	// Output: 2
}

func ExamplePushBubblesUp() {
	pq := New()
	pq.Push(3, "hello")
	pq.Push(2, "world")
	fmt.Println(pq.Peek())
	// Output: world <nil>
}

func ExamplePushBubblesUp2() {
	pq := New()
	pq.Push(3, "hello")
	pq.Push(2, "world")
	pq.Push(5, "dag")
	pq.Push(1, "wereld")
	pq.Push(10, "hoi")
	fmt.Println(pq.Peek())
	// Output: wereld <nil>
}

func ExamplePopBubblesDown() {
	pq := New()
	pq.Push(3, "hello")
	pq.Push(2, "world")
	pq.Push(5, "dag")
	pq.Push(1, "wereld")
	pq.Push(10, "oom")
	pq.Pop()
	fmt.Println(pq.Peek())
	// Output: world <nil>
}

func ExamplePopBubblesDown2() {
	pq := New()
	pq.Push(20, "word")
	pq.Push(0, "test")
	pq.Push(3, "hello")
	pq.Push(2, "world")
	pq.Push(5, "dag")
	pq.Push(1, "wereld")
	pq.Push(10, "oom")

	fmt.Println(pq.Pop())
	fmt.Println(pq.Pop())
	fmt.Println(pq.Pop())
	fmt.Println(pq.Pop())
	fmt.Println(pq.Pop())
	fmt.Println(pq.Pop())
	fmt.Println(pq.Pop())

	// Output:
	// test <nil>
	// wereld <nil>
	// world <nil>
	// hello <nil>
	// dag <nil>
	// oom <nil>
	// word <nil>
}

func ExamplePopOnEmptyError() {
	pq := New()
	_, error := pq.Pop()
	fmt.Println(error.Error())
	// Output: Queue is empty
}

func ExamplePeekOnEmptyError() {
	pq := New()
	_, error := pq.Peek()
	fmt.Println(error.Error())
	// Output: Queue is empty
}
